bigtable-dataflow-parent/bigtable-hbase-beam/src/main/java/com/google/cloud/bigtable/beam/CloudBigtableScanConfiguration.java

```diff
@@ -164,17 +229,67 @@ public CloudBigtableScanConfiguration build() {
           scan = new Scan();
         }
         ReadRowsRequest.Builder builder = Adapters.SCAN_ADAPTER.adapt(scan, readHooks);
-        request = readHooks.applyPreSendHook(builder.build());
+        request =  StaticValueProvider.of(readHooks.applyPreSendHook(builder.build()));
       }
       return new CloudBigtableScanConfiguration(projectId, instanceId, tableId,
           request, additionalConfiguration);
     }
   }
 
-  // 'request' needs to be a runtime parameter because it depends on runtime parameters: project ID
-  // and instance ID.
   private final ValueProvider<ReadRowsRequest> request;
 
+  /**
+   * Provides an updated request by setting the table name in the existing request if the table name
+   * wasn't set.
+   */
+  private static class RequestWithTableNameValueProvider
+      implements ValueProvider<ReadRowsRequest>, Serializable {
+    private final ValueProvider<String> projectId;
+    private final ValueProvider<String> instanceId;
+    private final ValueProvider<String> tableId;
+    private final ValueProvider<ReadRowsRequest> request;
+    private transient volatile ReadRowsRequest cachedRequest;
+
+    RequestWithTableNameValueProvider(
+        ValueProvider<String> projectId,
+        ValueProvider<String> instanceId,
+        ValueProvider<String> tableId,
+        ValueProvider<ReadRowsRequest> request) {
+      this.projectId = projectId;
+      this.instanceId = instanceId;
+      this.tableId = tableId;
+      this.request = request;
+    }
+
+    @Override
+    public ReadRowsRequest get() {
+      if (cachedRequest == null) {
+        if (request.get().getTableName().isEmpty()) {
+          BigtableInstanceName bigtableInstanceName =
+              new BigtableInstanceName(projectId.get(), instanceId.get());
+          String fullTableName = bigtableInstanceName.toTableNameStr(tableId.get());
+          cachedRequest = request.get().toBuilder().setTableName(fullTableName).build();
+        } else {
+          cachedRequest = request.get();
+        }
+      }
+      return cachedRequest;
+    }
+
+    @Override
+    public boolean isAccessible() {
+      return projectId.isAccessible()
+          && instanceId.isAccessible()
+          && tableId.isAccessible()
+          && request.isAccessible();
+    }
+
+    @Override
+    public String toString() {
+      return String.valueOf(get());
```

Comment:
having toString throw an exception makes for a really bad debugging experience

Reply:
Update to check isAccessible(), if not accessible, returns "Unavailable during pipeline construction";